### DISCUSSION TITLE
How window state restoring logic works?

### DISCUSSION THREAD

**Discussion by gnuarm** [sumatrapdfreader/sumatrapdf#2324] [Documentation]
I'm not clear on the various ways Sumatra can be shut down and the impact on how many of the windows open when the app is shut down are reopened when the app is started again. 

Seems you can use : 

File-Exit (Cntl-Q)
In Windows Taskbar, right-click then select Close all windows
From Task Manager select End Task
App crash or log out Windows (same thing, or not?) 

Any that I missed?  How many of these are equivalent from the point of view of the app? 

I would prefer to see all windows open at the time Sumatra was previously closed to be opened when Sumatra is opened again, regardless of the means of closing.

**Comment by kjk** [comment]
Background information:
* on Windows, the app exits when last window is closed
* we save the state on app exit

When only one window is opened, only one window is restored.

When multiple windows are opened, there are 2 scenarios:
a) multiple windows are restored
b) one window is restored

**When multiple windows are restored?** 

File/Exit (Ctrl-Q), app crash or forcibly killing the app.

**When only one window is restored?**

 All other cases.

**Why only one?** 

Because if you do Alt-F4 or Ctrl-W or use window close button to close a window, then you're just closing the window, not existing the app.

The state is saved when you quit the app i.e. close last window.

At that time the state of the app is just one window so that is the state restored.

End Task, Close all windows (and probably log out) do the equivalent of closing a window one by one, so that's why only one windows is restored.

**Reply by gnuarm** [reply]
> When multiple windows are restored?
> File/Exit (Ctrl-Q), app crash or forcibly killing the app.

So what does killing the program from Windows Task Manager do?  Does that save all the windows, or just the last one?  How about when the PC is rebooted? 

I ask because I normally kill apps in the Task Manager, which seems to allow most programs to recover their previous state, prior to being killed.  Sometimes, I need to reboot the computer to install software, then I usually let various programs be killed by rebooting the PC.  

I just manually killed Sumatra using Task Manager to reboot the PC and it seems to have recovered only the one window.  I thought this is what I've always done and had it reopen all my windows.  v3.4.14267 64-bit prerelease.

**Comment by GitHubRulesOK** [comment]
Worth adding that 
**When multiple windows are restored?**
when doing daily update open windows are reopened and restored to state prior to installation
